File: atcoder/abc/abc353/g/main.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

template<typename Monoid>
struct segtree{
  using F = function< Monoid(Monoid, Monoid) >;
  int size;
  const Monoid e;
  const F op;
  vector<Monoid> seg;
  segtree(int _N, F _op, const Monoid &_e): op(_op), e(_e){
    size = 1;
    while(size<_N) size <<= 1;
    seg.assign(2*size, e);
  }

  Monoid fold(int l, int r){
    Monoid vl = e, vr = e;
    for(l += size, r += size; l < r; l >>= 1, r >>= 1){
      if(l&1)vl = op(vl, seg[l++]);
      if(r&1)vr = op(vr, seg[--r]);
    }
    return op(vl, vr);
  }

  void set(int i, Monoid x){
    i += size;
    seg[i] = op(x, seg[i]);
    while(i>>=1){
      seg[i] = op(seg[i*2], seg[i*2+1]);
    }
  }

  Monoid get(int i){
    return seg[i+size];
  }

  void debug(){
    for(int i=0;i<size;i++){
      cout << seg[i+size] << ' ';
    }
    cout << endl;
  }
};

const ll INF = 1e18;
int main(){
  int N;
  ll C;
  cin >> N >> C;
  int M;
  cin >> M;
  ll T, P;
  segtree<ll> seg1(N, [](ll a, ll b){return max(a, b);}, -INF);
  segtree<ll> seg2(N, [](ll a, ll b){return max(a, b);}, -INF);
  seg1.set(0, 0);
  seg2.set(0, 0);
  ll l, r, v;

  for(int i=0;i<M;i++){
    cin >> T >> P;
    T--;
    l = seg1.fold(T, N) + C*T + P;
    r = seg2.fold(0, T) - C*T + P;
    v = max(l, r);
    seg1.set(T, v - C*T);
    seg2.set(T, v + C*T);
  }

  ll ans = 0;
  for(int i=0;i<N;i++){
    ans = max(ans, seg1.get(i) + i*C);
  }
  cout << ans << endl;
}

File: atcoder/abc/abc353/g/main.py

class SegTree:
	#in the implementation of segment tree its 1-indexed 
	#but in use case, it is 0-indexed
	def __init__(self, N, func, e=0):
		self._e = e
		self._func = func
		self._n = 1<<(N.bit_length())
		self.array = [e]*(self._n*2)

	def build(self, seq):
		for i, x in enumerate(seq):
			self.array[i+self._n] = x
		for i in range(1, self._n)[::-1]:
			self.array[i] = self._func(self.array[2*i], self.array[2*i+1])
	
	def update(self, i, x):
		i += self._n
		self.array[i] = x
		while i>1:
			i //= 2
			self.array[i] = self._func(self.array[2*i], self.array[2*i+1])
	
	def fold(self, l, r):
		l += self._n
		r += self._n
		vl = self._e
		vr = self._e
		while r-l>0:
			if r&1:
				r -= 1
				vr = self._func(vr, self.array[r])
			if l&1:
				vl = self._func(vl, self.array[l])
				l += 1
			r //= 2
			l //= 2
		return self._func(vl, vr)

	def all_fold(self):
		return self.array[1]
	
	def get(self, i):
		return self.array[i+self._n]

N, C = map(int, input().split())
INF  = 10**18
M = int(input())
seg1 = SegTree(N, max, -INF)
seg2 = SegTree(N, max, -INF)
seg1.update(0, 0)
seg2.update(0, 0)
for i in range(M):
	T, P = map(int, input().split())
	T -= 1
	l = seg1.fold(T, N) + T*C + P
	r = seg2.fold(0, T) - T*C + P
	v = max(l, r)
	seg1.update(T, v - T*C)
	seg2.update(T, v + T*C)
ans = 0
for i in range(N):
	ans = max(seg1.get(i) + i*C, ans)

print(ans)